In [30]:
import pandas as pd

# Reemplaza 'ruta/a/tu/archivo.xlsx' con la ruta real de tu archivo
ruta_archivo_excel = '/home/jp/Repositorios/VolumetricAnalysis_MagneticResonance/Data/Datos demográficos de todas las bases.xlsx'
nombre_hoja = 'Biomarcadores SOLO portadores y'

# Lee la hoja específica del archivo Excel
df = pd.read_excel(ruta_archivo_excel, sheet_name=nombre_hoja)
# Asumiendo que tu DataFrame se llama 'df'

# Selecciona las primeras 60 filas (índices 0 a 59)
df = df.iloc[:61]

# Opcional: si quieres modificar el DataFrame original directamente
# df = df.iloc[:60]

# Para verificar, puedes imprimir la forma (shape) del nuevo DataFrame
print(f"Dimensiones del DataFrame filtrado: {df.shape}")
display (df)

Dimensiones del DataFrame filtrado: (61, 39)


Code Biomarkers code  verificacion  Entry code Change of group  \
0   SAN_001         SAN_001            1.0        SAN              NO   
1   SAN_003         SAN_003            1.0        SAN              NO   
2   SAN_004         SAN_004            1.0        SAN              NO   
3   SAN_005         SAN_005            1.0        SAN              NO   
4   SAN_006         SAN_006            1.0        SAN              NO   
..      ...             ...            ...        ...             ...   
56     S096         SAN_091            1.0        SAN              NO   
57     S097         SAN_092            1.0        SAN              NO   
58     S095         SAN_093            1.0        SAN              NO   
59     S108         SAN_094            1.0        SAN              NO   
60     S104         SAN_095            1.0        SAN              NO   

   Exclusion Source project Medical reason  \
0         NO  Biomarcadores            NaN   
1         NO  Biomarcadores            NaN   
2         NO  Biomarcadores            NaN   
3         NO  Biomarcadores            NaN   
4         NO  Biomarcadores            NaN   
..       ...            ...            ...   
56       NaN          Suiza            NaN   
57        NO          Suiza            NaN   
58       NaN          Suiza            NaN   
59       NaN          Suiza            NaN   
60        NO          Suiza            NaN   

                              Subject name            ID  ...   T1   T2  DTI  \
0         Blanca Carolina Cuartas Martinez  1.216718e+09  ...    x    x    x   
1                Ana Maria Betancur Zapata  1.017218e+09  ...    x    x    x   
2                Andrey Felipe López López  1.042768e+09  ...    x    x    x   
3            Jennyfer Natalia Perez Balbin  1.152702e+09  ...    x    x    x   
4         Jessica Yesenia Cardenas Londoño  1.020433e+09  ...    x    x    x   
..                                     ...           ...  ...  ...  ...  ...   
56            SANDRA MILENA BETANCUR VILLA  3.256296e+07  ...  NaN  NaN  NaN   
57         BEATRIZ ELENA ORREGO PIEDRAHITA  4.382221e+07  ...  NaN  NaN  NaN   
58            ORFA ENITH MACIAS PIEDRAHITA  4.382228e+07  ...  NaN  NaN  NaN   
59     MARIA ALEXANDRA SAMPEDRO PIEDRAHITA  4.388382e+07  ...  NaN  NaN  NaN   
60  NATALIA ANDREA LONDOÃ_x0091_O SAMPEDRO  4.318470e+07  ...  NaN  NaN  NaN   

    SWI swimap          fMRI  Column 36 Column 37  Column 38 Column 39  
0     x      x   Run1 y Run2        NaN       NaN        NaN       NaN  
1     x      x  Orig y modif        NaN       NaN        NaN       NaN  
2     x      x  Orig y modif        NaN       NaN        NaN       NaN  
3     x      x   Run1 y Run2        NaN       NaN        NaN       NaN  
4     x      x   Run1 y Run2        NaN       NaN        NaN       NaN  
..  ...    ...           ...        ...       ...        ...       ...  
56  NaN    NaN           NaN        NaN       NaN        NaN       NaN  
57  NaN    NaN           NaN        NaN       NaN        NaN       NaN  
58  NaN    NaN           NaN        NaN       NaN        NaN       NaN  
59  NaN    NaN           NaN        NaN       NaN        NaN       NaN  
60  NaN    NaN           NaN        NaN       NaN        NaN       NaN  

[61 rows x 39 columns]

In [31]:
import pandas as pd

# Asumiendo que tu DataFrame original se llama 'df'
# y que ya lo has cargado previamente

# Lista con los nombres de las columnas que quieres CONSERVAR
columnas_a_conservar = ['Code','Biomarkers code','Birthdate', 'Sex', 'E280A group', 'MRI1 date'] # Reemplaza con los nombres reales

# Verifica que todas las columnas a conservar existan en el DataFrame original
columnas_faltantes = [col for col in columnas_a_conservar if col not in df.columns]

if columnas_faltantes:
    print(f"Error: Las siguientes columnas especificadas para conservar no existen en el DataFrame: {columnas_faltantes}")
    # Opcional: podrías detener la ejecución o manejar el error como prefieras
    # Por ejemplo, podrías conservar solo las que sí existen:
    # columnas_a_conservar = [col for col in columnas_a_conservar if col in df.columns]
    # print(f"Se conservarán solo las siguientes columnas existentes: {columnas_a_conservar}")
else:
    # Crea un nuevo DataFrame 'df_filtrado' que contiene solo las columnas especificadas
    # (O puedes sobreescribir el original usando: df = df[columnas_a_conservar])
    df_filtrado = df[columnas_a_conservar]

    # --- Verificación ---
    print(f"\nDataFrame 'df_filtrado' creado/actualizado.")
    print("Columnas conservadas:")
    print(df_filtrado.columns)
    print("\nPrimeras filas del DataFrame con columnas seleccionadas:")
    print(df_filtrado.head())



DataFrame 'df_filtrado' creado/actualizado.
Columnas conservadas:
Index(['Code', 'Biomarkers code', 'Birthdate', 'Sex', 'E280A group',
       'MRI1 date'],
      dtype='object')

Primeras filas del DataFrame con columnas seleccionadas:
      Code Biomarkers code            Birthdate Sex E280A group  MRI1 date
0  SAN_001         SAN_001  1995-07-26 00:00:00   F          GW 2023-04-14
1  SAN_003         SAN_003  1994-04-18 00:00:00   F          GK 2022-12-14
2  SAN_004         SAN_004  1990-08-29 00:00:00   M          GW 2023-01-10
3  SAN_005         SAN_005  1996-02-01 00:00:00   F          GK 2022-12-07
4  SAN_006         SAN_006  1990-07-25 00:00:00   F          GK 2023-01-28


In [32]:
# Asumiendo que tu DataFrame se llama df_filtrado
# y que contiene las columnas 'Code' y 'Birthdate'

# Identificador del paciente y la nueva fecha de nacimiento
paciente_id = 'S067'
nueva_fecha_nacimiento_str = '1980-07-21' # Usar formato YYYY-MM-DD o uno que pandas entienda

# Convertir la nueva fecha a formato datetime
nueva_fecha_nacimiento = pd.to_datetime(nueva_fecha_nacimiento_str)

# Verificar si el paciente existe en el DataFrame
if paciente_id in df_filtrado['Code'].values:
    # Usar .loc para encontrar la fila por el 'Code' y asignar la nueva fecha a 'Birthdate'
    df_filtrado.loc[df_filtrado['Code'] == paciente_id, 'Birthdate'] = nueva_fecha_nacimiento
    print(f"Fecha de nacimiento actualizada para el paciente {paciente_id}.")

    # --- Verificación ---
    print("\nFila actualizada:")
    print(df_filtrado[df_filtrado['Code'] == paciente_id])

    # Recalcular la edad para este paciente si ya existe la columna 'Age at MRI'
    if 'Age at MRI' in df_filtrado.columns:
        # Asegurarse de que 'MRI1 date' también sea datetime
        df_filtrado['MRI1 date'] = pd.to_datetime(df_filtrado['MRI1 date'], errors='coerce')
        
        # Obtener las fechas para el paciente específico
        birthdate = df_filtrado.loc[df_filtrado['Code'] == paciente_id, 'Birthdate'].iloc[0]
        mri_date = df_filtrado.loc[df_filtrado['Code'] == paciente_id, 'MRI1 date'].iloc[0]

        # Calcular la diferencia y la edad solo si ambas fechas son válidas
        if pd.notna(birthdate) and pd.notna(mri_date):
            diferencia = mri_date - birthdate
            edad = diferencia.days / 365.25
            df_filtrado.loc[df_filtrado['Code'] == paciente_id, 'Age at MRI'] = edad
            print(f"Edad recalculada para {paciente_id}: {edad:.2f} años.")
        else:
             df_filtrado.loc[df_filtrado['Code'] == paciente_id, 'Age at MRI'] = pd.NA # O np.nan si prefieres
             print(f"No se pudo recalcular la edad para {paciente_id} debido a fechas faltantes/inválidas.")

else:
    print(f"Error: No se encontró al paciente con Code '{paciente_id}' en el DataFrame.")

Fecha de nacimiento actualizada para el paciente S067.

Fila actualizada:
    Code Biomarkers code            Birthdate Sex E280A group  MRI1 date
51  S067         SAN_086  1980-07-21 00:00:00   F          GW 2023-04-28


In [40]:
import pandas as pd

# Asumiendo que tu DataFrame se llama df_filtrado

# 1. Asegurarse de que las columnas de fecha son de tipo datetime
#    errors='coerce' convertirá fechas inválidas a NaT (Not a Time)
df_filtrado['Birthdate'] = pd.to_datetime(df_filtrado['Birthdate'], errors='coerce')
df_filtrado['MRI1 date'] = pd.to_datetime(df_filtrado['MRI1 date'], errors='coerce')

# Opcional: Verificar si hubo errores en la conversión
if df_filtrado['Birthdate'].isnull().any() or df_filtrado['MRI1 date'].isnull().any():
    print("Advertencia: Algunas fechas no pudieron ser convertidas. Esas filas tendrán edad NaN.")
    # print("Filas con fechas inválidas:")
    # print(df_filtrado[df_filtrado['Birthdate'].isnull() | df_filtrado['MRI1 date'].isnull()])

# 2. Calcular la diferencia de tiempo entre las dos fechas
diferencia_tiempo = df_filtrado['MRI1 date'] - df_filtrado['Birthdate']

# 3. Crear la nueva columna 'Age at MRI' convirtiendo la diferencia a años
#    Dividimos los días totales por 365.25 para una aproximación que considera años bisiestos
df_filtrado['Age at MRI'] = diferencia_tiempo.dt.days / 365.25




/tmp/ipykernel_103226/2019072451.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtrado['Birthdate'] = pd.to_datetime(df_filtrado['Birthdate'], errors='coerce')
/tmp/ipykernel_103226/2019072451.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtrado['MRI1 date'] = pd.to_datetime(df_filtrado['MRI1 date'], errors='coerce')
/tmp/ipykernel_103226/2019072451.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] 

In [34]:
df_filtrado

Code Biomarkers code  Birthdate Sex E280A group  MRI1 date  Age at MRI
0   SAN_001         SAN_001 1995-07-26   F          GW 2023-04-14   27.718001
1   SAN_003         SAN_003 1994-04-18   F          GK 2022-12-14   28.657084
2   SAN_004         SAN_004 1990-08-29   M          GW 2023-01-10   32.366872
3   SAN_005         SAN_005 1996-02-01   F          GK 2022-12-07   26.847365
4   SAN_006         SAN_006 1990-07-25   F          GK 2023-01-28   32.511978
..      ...             ...        ...  ..         ...        ...         ...
56     S096         SAN_091 1981-12-24   F          GW 2023-07-05   41.527721
57     S097         SAN_092 1980-03-17   F          GW 2023-07-05   43.299110
58     S095         SAN_093 1978-09-10   F          GW 2023-07-04   44.813142
59     S108         SAN_094 1980-09-07   F          GK 2023-07-19   42.861054
60     S104         SAN_095 1983-10-16   F          GW 2023-07-14   39.742642

[61 rows x 7 columns]

In [ ]:
import pandas as pd
from scipy import stats
from IPython.display import display # To display the table nicely in Jupyter

# Asumiendo que tu DataFrame se llama df_filtrado

# 1. Filtrar solo los grupos GK y GW
df_gw_gk = df_filtrado[df_filtrado['E280A group'].isin(['GK', 'GW'])].copy()

# Verificar si tenemos datos para ambos grupos
if 'GK' not in df_gw_gk['E280A group'].unique() or 'GW' not in df_gw_gk['E280A group'].unique():
    print("Error: No se encontraron datos para ambos grupos 'GK' y 'GW' en el DataFrame.")
# Verificar si el dataframe filtrado está vacío
elif df_gw_gk.empty:
    print("Error: El DataFrame filtrado para los grupos 'GK' y 'GW' está vacío.")
else:
    # --- Cálculos ---

    # a) Edad ('Age at MRI') - Media (Desviación Estándar)
    age_summary = df_gw_gk.groupby('E280A group')['Age at MRI'].agg(['mean', 'std', 'count'])
    # Asegurarse que ambos grupos existen en el summary
    if 'GK' in age_summary.index and 'GW' in age_summary.index:
        age_gk_str = f"{age_summary.loc['GK', 'mean']:.2f} ({age_summary.loc['GK', 'std']:.2f})"
        age_gw_str = f"{age_summary.loc['GW', 'mean']:.2f} ({age_summary.loc['GW', 'std']:.2f})"
        n_gk = int(age_summary.loc['GK', 'count'])
        n_gw = int(age_summary.loc['GW', 'count'])

        # Prueba T para comparar edad entre GK y GW
        gk_ages = df_gw_gk[df_gw_gk['E280A group'] == 'GK']['Age at MRI'].dropna()
        gw_ages = df_gw_gk[df_gw_gk['E280A group'] == 'GW']['Age at MRI'].dropna()
        if len(gk_ages) > 1 and len(gw_ages) > 1:
            t_stat_age, p_val_age = stats.ttest_ind(gk_ages, gw_ages)
            p_val_age_str = f"{p_val_age:.3f}"
            if p_val_age < 0.001: p_val_age_str = "<0.001"
        else:
            p_val_age_str = "N/A" # Not enough data
    else:
        print("Error: Falta uno o ambos grupos ('GK', 'GW') después de agrupar por edad.")
        age_gk_str, age_gw_str, n_gk, n_gw, p_val_age_str = "Error", "Error", 0, 0, "Error"


    # b) Sexo - N (%)
    # Verificar si hay datos de sexo antes de agrupar
    if not df_gw_gk[['E280A group', 'Sex']].dropna().empty:
        sex_counts = df_gw_gk.groupby('E280A group')['Sex'].value_counts().unstack(fill_value=0)
        # Asegurarse de que existan las columnas M y F incluso si no hay datos para un sexo
        if 'M' not in sex_counts.columns: sex_counts['M'] = 0
        if 'F' not in sex_counts.columns: sex_counts['F'] = 0
        sex_counts = sex_counts[['M', 'F']] # Asegurar orden

        sex_totals = sex_counts.sum(axis=1)
        # Evitar división por cero si un grupo no tiene datos de sexo
        sex_percentages = sex_counts.apply(lambda x: (x / sex_totals[x.name]) * 100 if sex_totals[x.name] > 0 else 0, axis=1)

        # Asegurarse que ambos grupos existen en los conteos
        male_gk_str = f"{sex_counts.loc['GK', 'M']} ({sex_percentages.loc['GK', 'M']:.1f})" if 'GK' in sex_counts.index else "N/A"
        female_gk_str = f"{sex_counts.loc['GK', 'F']} ({sex_percentages.loc['GK', 'F']:.1f})" if 'GK' in sex_counts.index else "N/A"
        male_gw_str = f"{sex_counts.loc['GW', 'M']} ({sex_percentages.loc['GW', 'M']:.1f})" if 'GW' in sex_counts.index else "N/A"
        female_gw_str = f"{sex_counts.loc['GW', 'F']} ({sex_percentages.loc['GW', 'F']:.1f})" if 'GW' in sex_counts.index else "N/A"

        # Prueba de Fisher (más adecuada para tablas 2x2, como Sexo vs Grupo)
        # Crear tabla de contingencia solo con datos no nulos
        contingency_table_sex = pd.crosstab(df_gw_gk['E280A group'], df_gw_gk['Sex'])
        
        # Verificar si la tabla es válida (2x2) para Fisher
        if contingency_table_sex.shape == (2, 2) and contingency_table_sex.notna().all().all():
            oddsr, p_val_sex = stats.fisher_exact(contingency_table_sex)
            p_val_sex_str = f"{p_val_sex:.3f}"
            # Si p-valor es muy pequeño, mostrar "<0.001"
            if p_val_sex < 0.001: p_val_sex_str = "<0.001"
        else:
             # Intentar Chi-cuadrado si Fisher no aplica (e.g., tabla no 2x2 pero con suficientes datos)
            try:
                chi2, p_val_sex_chi, dof, expected = stats.chi2_contingency(contingency_table_sex)
                p_val_sex_str = f"{p_val_sex_chi:.3f} (Chi2)"
                if p_val_sex_chi < 0.001: p_val_sex_str = "<0.001 (Chi2)"
            except ValueError: # Pocos datos para Chi2
                p_val_sex_str = "N/A" # No se puede calcular test

    else:
        print("Advertencia: No hay datos de sexo disponibles para calcular la distribución.")
        male_gk_str, female_gk_str, male_gw_str, female_gw_str, p_val_sex_str = "N/A", "N/A", "N/A", "N/A", "N/A"


    # --- Crear la tabla DataFrame ---
    # Solo crear tabla si n_gk y n_gw son válidos (mayores que 0)
    if n_gk > 0 and n_gw > 0 :
        table_data = {
            'Variable': [
                f'Age - Mean (SD)',
                'Sex - N (%)',
                '  Male',
                '  Female'
            ],
            f'GK (n={n_gk})': [
                age_gk_str,
                '', # Fila de encabezado para Sexo
                male_gk_str,
                female_gk_str
            ],
            f'GW (n={n_gw})': [
                age_gw_str,
                '', # Fila de encabezado para Sexo
                male_gw_str,
                female_gw_str
            ],
            'p-value (GK vs GW)': [
                p_val_age_str,
                '', # Fila de encabezado para Sexo
                '', # p-value general para Sexo se muestra en la siguiente línea
                p_val_sex_str
            ]
        }

        summary_table = pd.DataFrame(table_data)
        summary_table.set_index('Variable', inplace=True)

        # --- Mostrar la tabla ---
        print("\n--- Summary Table (GK vs GW) ---")
        display(summary_table) # Muestra la tabla formateada en el notebook

        print("\nNotes:")
        print("- Age is presented as Mean (Standard Deviation).")
        print("- Sex is presented as Count (Percentage).")
        print("- P-values compare GK vs GW groups using T-test for Age and Fisher's Exact Test (or Chi2) for Sex.")
        print("- Variables like Education, MMSE, CDR are not included as they were not found in the provided DataFrame.")
    else:
        print("\nNo se pudo generar la tabla resumen porque uno o ambos grupos (GK, GW) no tienen datos suficientes.")

print(summary_table.to_string())



--- Summary Table (GK vs GW) ---


GK (n=25)     GW (n=36) p-value (GK vs GW)
Variable                                                      
Age - Mean (SD)  36.01 (6.66)  35.68 (7.24)              0.857
Sex - N (%)                                                   
  Male               9 (36.0)     11 (30.6)                   
  Female            16 (64.0)     25 (69.4)              0.783


Notes:
- Age is presented as Mean (Standard Deviation).
- Sex is presented as Count (Percentage).
- P-values compare GK vs GW groups using T-test for Age and Fisher's Exact Test (or Chi2) for Sex.
- Variables like Education, MMSE, CDR are not included as they were not found in the provided DataFrame.
                    GK (n=25)     GW (n=36) p-value (GK vs GW)
Variable                                                      
Age - Mean (SD)  36.01 (6.66)  35.68 (7.24)              0.857
Sex - N (%)                                                   
  Male               9 (36.0)     11 (30.6)                   
  Female            16 (64.0)     25 (69.4)              0.783


| Variable         | GK (n=25)    | GW (n=35)    | p-value (GK vs GW)   |
|:-----------------|:-------------|:-------------|:---------------------|
| Age - Mean (SD)  | 36.01 (6.66) | 35.56 (7.31) | 0.810                |
|   Male           | 9 (36.0)     | 11 (31.4)    |                      |
|   Female         | 16 (64.0)    | 24 (68.6)    | 0.785                |




In [9]:
import pandas as pd
import os # Importar os para manejar rutas

output_directory = 'Demographic_data' # Cambia esto
# --- Asegúrate de que df_filtrado existe ---
if 'df_filtrado' in locals() and isinstance(df_filtrado, pd.DataFrame):

    output_filename = 'Demographic_data.csv' # Cambia esto
    full_output_path = os.path.join(output_directory, output_filename)

    try:
        # Crear el directorio si no existe
        os.makedirs(output_directory, exist_ok=True) # exist_ok=True evita error si ya existe

        # --- Guardar el DataFrame en CSV ---
        # index=False evita que el índice del DataFrame se escriba como una columna en el CSV
        df_filtrado.to_csv(full_output_path, index=False, encoding='utf-8')

        print(f"DataFrame 'df_filtrado' guardado exitosamente en:")
        print(f"{os.path.abspath(full_output_path)}") # Muestra la ruta absoluta

    except Exception as e:
        print(f"Error al guardar el archivo CSV en '{full_output_path}':")
        print(e)

else:
    print("Error: El DataFrame 'df_filtrado' no existe o no es válido.")
    print("Asegúrate de que se haya creado correctamente en una celda anterior.")


DataFrame 'df_filtrado' guardado exitosamente en:
/home/jp/Repositorios/VolumetricAnalysis_MagneticResonance/Notebooks/Demographic_data/Demographic_data.csv
